### ISSUE TITLE
Proper dma read card functions implementations

### ISSUE THREAD

**Issue by ahezard** [DS-Homebrew/nds-bootstrap#517]
The current implementation of the dma read card functions deactivate all dma card read. This behavior can affect performance and compatibility.

Game that uses dma to acces the card :
GTA chinatown wars
Golden Sun
Elektroplanton
Castlevania Dawn of Sorrow
Pokemon Dash
Kirby Canvas Curse
Mario & Luigi: Bowser's Inside Story (sometimes)
Spectrobes
Professor Layton and the Curious Village (during cutscenes)
Meteos
Final Fantasy XII - Revenant Wings
Maplestory
Mr Bean
Yu-Gi-Oh 5Ds World Championship 2011: Over The Nexus
Assassin's Creed - Altair's Chronicles
Nanashi no Geemu
Theme Park
Tales of Innocence

How to spot a game that uses dma via no$gba : 
Enable DMA log on ARM9
A DMA access to the card uses AF000001 as the third parameter
eg : DMA2: 04100010 023C18C0 AF000001

**Comment by MasterGamingYT** [comment]
oof

**Comment by ahezard** [comment]
As of now this full method is available on master. Some more testing is needed.

**Comment by hubert-skowronek** [comment]
@ahezard  Tested 6be5c8a:
Mario & Luigi BiS - graphical glitches during cutscenes still present (horizontal bars flickering)
Hotel Dusk - green bars flickering during cutscenes still present

**Comment by ahezard** [comment]
6be5c8a did not have #517 active for Mario & Luigi BiS please retest with 2e48916

**Comment by hubert-skowronek** [comment]
I have a black screen for 2e48916 when running ML BiS

**Comment by hubert-skowronek** [comment]
@ahezard with 9de36d8 ML BiS loads but after choosing save it is the black screen

**Comment by NightYoshi370** [comment]
Hasn't this been implemented with the release v0.29.0?

**Comment by ahezard** [comment]
Yes it have been implemented but I consider the work to not be 100% done. Implementation for SDK5 games still have issues. I am working on it.

**Comment by RocketRobz** [comment]
As of recent commits, DMA reads aren't working with heap shrink turned on.
I have no idea why that is, as it worked just fine before with heap shrink.

**Comment by luigiandmariooooooo** [comment]
> Mario & Luigi BiS - graphical glitches during cutscenes still present (horizontal bars flickering)
Hotel Dusk - green bars flickering during cutscenes still present

I have noticed similar issues in luminous arc (horizontal black bars during cutscenes) and project rub (rare horizontal red bars during cutscenes), perhaps they are related to this whole dma read card function. Both games were EU versions and were played in NTR mode

my setup - new 2ds xl; twilight menu v15.1.1; nds bootstrap v0.35.1; luna v10.2

**Comment by RocketRobz** [comment]
As of recent commits, some games that rely on DMA reads, will either not boot, or crash at some point.
Examples are:
* Tony Hawk's American Sk8land
* Tony Hawk's Downhill Jam
* Yu-Gi-Oh 5Ds World Championship 2011: Over The Nexus

**Comment by zosxavius** [comment]
Looks like radiant Historia is having DMA issues again. At least on a japanese DSi LL. I get flickering in the title and at various points. Its happening during card reads looks like. I did play this before on an older version and I remember it being perfect like a year ago. That was with an American NDS. I doubt that matters. I have the 3ds version of radiant which is superior IMO, but just letting you know. I decided to boot up the DS version to compare a little. I checked and I'm on the latest NDS bootstrap release. It was perfect a year ago. I tried a donor ROM too and that didn't help. Turning DMA off makes it load slowly and doesn't fix anything.  I also had DQVI hang right after you leave the first town. I think that was on my 3ds. I'll have to test on the DSi XL. I don't remember that hanging before though. Unrelated probably but Mario Kart hangs on multiplayer unless running at 133mhz. 

There's some nice improvements here so I fully look forward to new releases.

**Comment by Kaisaan** [comment]
What games still have this issue on the latest version of nds-bootstrap?

**Comment by RocketRobz** [comment]
Only known one is Tropix, which crashes often with the DMA option turned on.

**Comment by Kaisaan** [comment]
Since turning off that setting fixes it, would it make this issue good to close?

**Comment by RocketRobz** [comment]
Yeah, our implementation seems to be mostly(?) good, at this point.

**Comment by RocketRobz** [comment]
Re-opening, since some unexpected regressions (mostly in DSi mode) have occurred since v0.63.0.
